File: AnalyticsKitFlurryProvider.m
//
//  AnalyticsKitFlurryProvider.m
//  TeamStream
//
//  Created by Susan Detwiler on 11/10/11.
//  Copyright (c) 2011 Bleacher Report. All rights reserved.
//

#import "Flurry.h"
#import "AnalyticsKitFlurryProvider.h"

@implementation AnalyticsKitFlurryProvider

-(id<AnalyticsKitProvider>)initWithAPIKey:(NSString *)apiKey {
    self = [super init];
    if (self) {
        [Flurry startSession:apiKey];
    }
    return self;
}

-(void)runInMainThread:(void (^)(void))block{
    //Flurry requires calls to be made from main thread
    if ([NSThread isMainThread]){
        block();
    } else {
        dispatch_async(dispatch_get_main_queue(), ^{
            block();
        });
    }
}

-(void)applicationWillEnterForeground {}
-(void)applicationDidEnterBackground {}
-(void)applicationWillTerminate {}

-(void)uncaughtException:(NSException *)exception {
    NSString *message = [NSString stringWithFormat:@"Crash on iOS %@", [[UIDevice currentDevice] systemVersion]];
    [Flurry logError:@"Uncaught" message:message exception:exception];
}

-(void)logScreen:(NSString *)screenName {
    [self runInMainThread:^{
        [Flurry logEvent:[@"Screen - " stringByAppendingString:screenName]];
    }];
}

-(void)logEvent:(NSString *)value {
    [self runInMainThread:^{
        [Flurry logEvent:value];
    }];
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict {
    [self runInMainThread:^{
        [Flurry logEvent:event withParameters:dict];
    }];
}

-(void)logEvent:(NSString *)event withProperty:(NSString *)key andValue:(NSString *)value {
    [self runInMainThread:^{
        [Flurry logEvent:event withParameters:[NSDictionary dictionaryWithObject:value forKey:key]];
    }];
}

- (void)logEvent:(NSString *)eventName timed:(BOOL)timed{
    [self runInMainThread:^{
        [Flurry logEvent:eventName timed:timed];
    }];
}

- (void)logEvent:(NSString *)eventName withProperties:(NSDictionary *)dict timed:(BOOL)timed{
    [self runInMainThread:^{
        [Flurry logEvent:eventName withParameters:dict timed:timed];
        
    }];
}

-(void)endTimedEvent:(NSString *)eventName withProperties:(NSDictionary *)dict{
    [self runInMainThread:^{
        // non-nil parameters will update the parameters
        [Flurry endTimedEvent:eventName withParameters:dict];
     }];
}

-(void)logError:(NSString *)name message:(NSString *)message exception:(NSException *)exception {
    [self runInMainThread:^{
        [Flurry logError:name message:message exception:exception];
    }];
}

-(void)logError:(NSString *)name message:(NSString *)message error:(NSError *)error {
    [self runInMainThread:^{
        [Flurry logError:name message:message error:error];
    }];
}

@end

File: Providers/Flurry/AnalyticsKitFlurryProvider.swift
import Foundation
import Flurry_iOS_SDK

public class AnalyticsKitFlurryProvider: NSObject, AnalyticsKitProvider {

    public init(withAPIKey apiKey: String) {
        Flurry.startSession(apiKey)
    }

    public func uncaughtException(_ exception: NSException) {
        let message = "Crash on iOS \(UIDevice.current.systemVersion)"
        Flurry.logError("Uncaught", message: message, exception: exception)
    }

    public func applicationWillEnterForeground() {}
    public func applicationDidEnterBackground() {}
    public func applicationWillTerminate() {}

    // Logging
    public func logScreen(_ screenName: String) {
        runOnMainThread {
            Flurry.logEvent("Screen - \(screenName)")
        }
    }

    public func logScreen(_ screenName: String, withProperties properties: [String: Any]) {
        runOnMainThread {
            Flurry.logEvent("Screen - \(screenName)", withParameters: properties)
        }
    }

    public func logEvent(_ event: String) {
        runOnMainThread { 
            Flurry.logEvent(event)
        }
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any]) {
        runOnMainThread {
            Flurry.logEvent(event, withParameters: properties)
        }
    }
    
    public func logEvent(_ event: String, withProperty key: String, andValue value: String) {
        logEvent(event, withProperties: [key: value])
    }

    public func logEvent(_ event: String, timed: Bool) {
        runOnMainThread {
            Flurry.logEvent(event, timed: timed)
        }
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any], timed: Bool) {
        runOnMainThread {
            Flurry.logEvent(event, withParameters: properties, timed: timed)
        }
    }

    public func endTimedEvent(_ event: String, withProperties properties: [String: Any]) {
        runOnMainThread {
            // non-nil parameters will update the parameters
            Flurry.endTimedEvent(event, withParameters: properties)
        }
    }

    public func logError(_ name: String, message: String?, exception: NSException?) {
        runOnMainThread { 
            Flurry.logError(name, message: message, exception: exception)
        }
    }

    public func logError(_ name: String, message: String?, error: Error?) {
        runOnMainThread {
            Flurry.logError(name, message: message, error: error)
        }
    }

    fileprivate func runOnMainThread(_ block: @escaping () -> Void) {
        //Flurry requires calls to be made from main thread
        if Thread.isMainThread {
            block()
        } else {
            DispatchQueue.main.async(execute: {
                block()
            })
        }
    }
}
